core/corehttp/gateway_handler.go

```diff
@@ -218,21 +210,74 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 }
 
 func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
-	nd, err := i.newDagFromReader(r.Body)
-	if err != nil {
-		internalWebError(w, err)
-		return
+	var resultHash string
+	var isDir bool
+	// if encoding is "tar/gzip", we assume its a directory
+	if v, ok := r.Header["Content-Encoding"]; ok && len(v) > 0 && v[0] == "tar/gzip" {
+		isDir = true
 	}
 
-	k, err := i.node.DAG.Add(nd)
-	if err != nil {
-		internalWebError(w, err)
-		return
-	}
+	if isDir {
+		// decompress, untar, and add directory to ipfs
+		tarBall, err := decompressBody(r)
+		if err != nil {
+			internalWebError(w, err)
+			return
+		}
+
+		rootDir, err := unarchiveTarBall(tarBall)
```

Comment:
yeah :/ -- maybe if they're adding a file with our tools, it would be wrapped for them again. but good point. maybe should have a way to say "this is exactly what i want"